components/brave_ads/core/internal/ad_units/creative_ad_cache.cc

```diff
@@ -0,0 +1,82 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/core/internal/ad_units/creative_ad_cache.h"
+
+#include <optional>
+
+#include "base/containers/contains.h"
+#include "base/functional/overloaded.h"
+#include "brave/components/brave_ads/core/internal/tabs/tab_info.h"
+#include "brave/components/brave_ads/core/internal/tabs/tab_manager.h"
+
+namespace brave_ads {
+
+namespace {
+
+std::optional<CreativeAdVariant> CloneCreativeAdEventInfo(
+    const CreativeAdVariant& creative_ad_event_info) {
+  return absl::visit(
+      base::Overloaded{
+          [&](const mojom::CreativeSearchResultAdInfoPtr& search_result_ad)
+              -> std::optional<CreativeAdVariant> {
+            return search_result_ad.Clone();
+          }},
+      creative_ad_event_info);
+}
+
+}  // namespace
+
+CreativeAdCache::CreativeAdCache() {
+  TabManager::GetInstance().AddObserver(this);
+}
+
+CreativeAdCache::~CreativeAdCache() {
+  TabManager::GetInstance().RemoveObserver(this);
+}
+
+void CreativeAdCache::Add(const std::string& placement_id,
+                          CreativeAdVariant creative_ad_variant) {
+  const std::optional<TabInfo> tab =
+      TabManager::GetInstance().MaybeGetVisible();
+  if (!tab) {
+    return;
+  }
+
+  creative_ad_variants_[placement_id] = std::move(creative_ad_variant);
+  placement_ids_[tab->id].push_back(placement_id);
+}
+
+std::optional<CreativeAdVariant> CreativeAdCache::MaybeGet(
+    const std::string& placement_id) {
+  const auto iter = creative_ad_variants_.find(placement_id);
+  if (iter == creative_ad_variants_.cend()) {
+    return std::nullopt;
+  }
+  const auto& creative_ad_variant = iter->second;
+
+  return CloneCreativeAdEventInfo(creative_ad_variant);
+}
+
+void CreativeAdCache::OnTabDidChange(const TabInfo& tab) {
+  //  PurgePlacements(tab.id);
```

Comment:
Is this intentional as if so, would be great to remove OnTabDidChange.

Reply:
Removed `OnTabDidChange`, per discussion in DM